File: apps/chat-android/messaging-websocket/src/main/java/com/amazonaws/services/chime/sdk/messaging/internal/DefaultWebSocketAdapter.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.amazonaws.services.chime.sdk.messaging.internal

import com.amazonaws.services.chime.sdk.messaging.session.MessagingSessionStatus
import com.amazonaws.services.chime.sdk.messaging.utils.logger.Logger
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

class DefaultWebSocketAdapter(
    private val logger: Logger
) : WebSocketAdapter {
        private val client: OkHttpClient =
        OkHttpClient.Builder()
            .pingInterval(60, TimeUnit.SECONDS)
            .build()
    private var session: WebSocket? = null

    private val TAG = "DefaultWebSocketAdapter"

    override fun create(url: String, observer: WebSocketAdapterObserver) {
        GlobalScope.launch {
            val request = Request.Builder().url(url).build()
            client.newWebSocket(request, object : WebSocketListener() {
                override fun onOpen(webSocket: WebSocket, response: Response) {
                    session = webSocket
                    logger.info(TAG, "onOpen response: $response")
                    observer.onConnect()
                }

                override fun onMessage(webSocket: WebSocket, text: String) {
                    observer.onMessage(text)
                }

                override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                    observer.onClose(MessagingSessionStatus(code, reason))
                }

                override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                    logger.error(TAG, "Connection failed: $response")
                    observer.onClose(MessagingSessionStatus(response?.code, response?.message))
                }

                override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                    observer.onClose(MessagingSessionStatus(code, reason))
                }
            })
        }
    }

    override fun close() {
        GlobalScope.launch {
            session?.close(1000, null)
            session = null
        }
    }
}

File: apps/chat-ios/AmazonChimeSDKMessagingDemo/messaging-websocket/internal/DefaultWebSocketAdapter.swift
//
//  DefaultWebSocketAdapter.swift
//  AmazonChimeSDKMessagingDemo
//
//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
//  SPDX-License-Identifier: MIT-0
//

import Starscream

public class DefaultWebSocketAdapter: WebSocketAdapter, WebSocketDelegate {
    private var socket: WebSocket?
    private var observer: WebSocketAdapterObserver?
    
    // MARK: - WebSocketAdapter
    public func connect(url: String, observer: WebSocketAdapterObserver) {
        self.observer = observer

        guard let urlObj = URL(string: url) else {
            print("DefaultWebSocketAdapter connect() bad url")
            return
        }
        var request = URLRequest(url: urlObj)
        request.timeoutInterval = 5
        socket = WebSocket(request: request)
        socket?.delegate = self
        socket?.connect()
    }

    public func close() {
        socket?.disconnect()
    }

    // MARK: - WebSocketDelegate
    public func didReceive(event: WebSocketEvent, client: WebSocket) {
        switch event {
            case .connected(let headers):
                print("DefaultWebSocketAdapter websocket is connected: \(headers)")
                observer?.onConnect()
            case .disconnected(let reason, let code):
                print("DefaultWebSocketAdapter websocket is disconnected: \(reason) with code: \(code)")
                observer?.onClose(status: MessagingSessionStatus(code: Int(code), reason: reason))
            case .text(let string):
                print("DefaultWebSocketAdapter text received: \(string)")
                observer?.onMessage(message: string)
            case .binary(let data):
                print("DefaultWebSocketAdapter binary received: \(data.count)")
            case .ping(_):
                print("DefaultWebSocketAdapter ping received")
                break
            case .pong(_):
                print("DefaultWebSocketAdapter pong received")
                break
            case .viabilityChanged(_):
                print("DefaultWebSocketAdapter viability changed")
                break
            case .reconnectSuggested(_):
                print("DefaultWebSocketAdapter reconnect suggested")
                break
            case .cancelled:
                print("DefaultWebSocketAdapter cancelled")
            case .error(let error):
                print("DefaultWebSocketAdapter error \(String(describing: error))")
        }
    }
}
